docs/grpc_tutorial.md

```diff
@@ -0,0 +1,265 @@
+This separation means that the learner can run on specialised hardware, e.g. a compute server, and does not need to be co-located with the backend.
+
+## Architecture of gRPC server
+The gRPC interface is defined in colearn_grpc/proto/interface.proto.
+This defines the functions that the gRPC server exposes and the format for messages between the server and the client.
+
+gRPC server has a mli factory, and it uses the mli factory to make the learner
+mli factory needs to implement mli factory interface.
+So you could write your own mli factory,
+but it's easier to use the one we provide
+Below we will discuss the mli factory and then talk about how to use the example factory.
+
+## MLI Factory interface:
+mli factory (as the name suggests) is a factory class for creating objects that implement the machine learning interface.
+mlif stores the constructors for dataloaders and models and also the dataloaders that are compatible with each model.
+Each construcotr is stored under a specific name.
+For example, "KERAS_MNIST_MODEL" is the model for keras mnist.
+The gRPC server uses the mli factory to construct mli objects.
+mlif needs to implement four methods:
+* get_models - returns the names of the models that are registered with the factory and their parameters
+* get_dataloaders - returns the names of the dataloaders that are registered with the factory and their parameters
+* get compatibilities - returns a list of dataloaders for each model that can be used with that model
+* get_mli - takes the name and parameters for the model and dataloader and contructs the mli object. Returns the mli object
+
+
+## Using the example MLI Factory
+The example mli factory is defined in colearn_grpc/example_mli_factory.py.
+It stores the models and dataloaders that it knows about in factoryRegistry.py
+To add a new model and dataloader to the factory you need to do the folowing things:
+1. Define a function that loads the dataset given the location of the dataset.
+2. Define a function that takes in the dataset and loads the mli model. 
+2. Register both these functions with the factory registry. 
+
+Registering a dataloader looks like this:
+```python
+@FactoryRegistry.register_dataloader(dataloader_tag)
+def prepare_data_loaders(train_folder: str,
+                         train_ratio: float = 0.9,
+                         batch_size: int = 32) -> Tuple[PrefetchDataset, PrefetchDataset]:
+```
+Registering a model is similar, but you additionally have to specify the dataloaders that this model is compatible with.
+```python
+@FactoryRegistry.register_model_architecture(model_tag, [dataloader_tag])
+def prepare_learner(data_loaders: Tuple[PrefetchDataset, PrefetchDataset],
+                    steps_per_epoch: int = 100,
+                    vote_batches: int = 10,
+                    learning_rate: float = 0.001
+                    ) -> KerasLearner:
+```
+
+You can see an example of how to do this in {{ repo_root }}/colearn_grpc/mnist_grpc.py.
```

Comment:
is this supposed to show as repo root?

Reply:
When the docs are built properly then it will be replaced with the address of the repo. I still need to make it a link though!